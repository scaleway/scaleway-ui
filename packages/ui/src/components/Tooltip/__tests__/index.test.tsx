import { act, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import type { ComponentProps } from 'react'
import { Tooltip } from '..'
import {
  renderWithTheme,
  shouldMatchEmotionSnapshot,
} from '../../../../.jest/helpers'

describe('Tooltip', () => {
  test('should render correctly', () =>
    shouldMatchEmotionSnapshot(<Tooltip text="test">Hover me</Tooltip>))

  test('should render correctly without text', () =>
    shouldMatchEmotionSnapshot(<Tooltip>Hover me</Tooltip>))

  test(`should display tooltip on hover`, async () => {
    const node = renderWithTheme(
      <Tooltip id="test" text="test success!">
        <p data-testid="children">Hover me</p>
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
  })

  test(`should display tooltip on hover with function children`, async () => {
    const node = renderWithTheme(
      <Tooltip id="test" text="test success!">
        {props => (
          <p {...props} data-testid="children">
            Hover me
          </p>
        )}
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
  })

  test(`should display tooltip on hover and hide when exit`, async () => {
    const node = renderWithTheme(
      <Tooltip id="test" text="test success!">
        <p data-testid="children">Hover me</p>
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
    await act(async () => {
      await userEvent.unhover(input)
    })
    await waitFor(() => {
      expect(tooltipPortal).not.toBeVisible()
    })
  })

  test(`should display tooltip on hover and hide when exit and hover back before animation ends`, async () => {
    const node = renderWithTheme(
      <Tooltip id="test" text="test success!">
        <p data-testid="children">Hover me</p>
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
    await act(async () => {
      await userEvent.unhover(input)
      await userEvent.hover(input)
    })
    expect(tooltipPortal).toBeVisible()
  })

  test(`should create tooltip with random id`, async () => {
    const node = renderWithTheme(
      <Tooltip text="test success!">
        <p data-testid="children">Hover me</p>
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
  })

  test(`should renders tooltip with maxWidth`, async () => {
    const node = renderWithTheme(
      <Tooltip text="test success!" maxWidth={100}>
        <p data-testid="children">Hover me</p>
      </Tooltip>,
    )

    const input = node.getByTestId('children') as HTMLInputElement
    await act(async () => {
      await userEvent.hover(input)
    })
    const tooltipPortal = node.getByText('test success!') as HTMLInputElement
    expect(tooltipPortal).toBeVisible()
  })

  describe(`defined placement`, () => {
    ;['top', 'left', 'right', 'bottom'].forEach(placement => {
      test(`should renders tooltip with placement ${placement}`, async () => {
        const node = renderWithTheme(
          <Tooltip
            text="test success!"
            placement={placement as ComponentProps<typeof Tooltip>['placement']}
          >
            <p data-testid="children">Hover me</p>
          </Tooltip>,
        )

        const input = node.getByTestId('children') as HTMLInputElement
        await act(async () => {
          await userEvent.hover(input)
        })
        const tooltipPortal = node.getByText(
          'test success!',
        ) as HTMLInputElement
        expect(tooltipPortal).toBeVisible()
      })
    })
  })
})
