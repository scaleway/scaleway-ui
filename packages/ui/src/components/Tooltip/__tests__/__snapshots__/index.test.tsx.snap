// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should render correctly 1`] = `
<DocumentFragment>
  .cache-yb56yg-StyledChildrenContainer {
  display: inherit;
}

<div
    aria-describedby=":r0:"
    class="cache-yb56yg-StyledChildrenContainer ep3agpf0"
    tabindex="0"
  >
    Hover me
  </div>
</DocumentFragment>
`;

exports[`Tooltip should render correctly without text 1`] = `
<DocumentFragment>
  Hover me
</DocumentFragment>
`;
